### PR TITLE
Pr add break build thresholds

### PR DESCRIPTION
**Pull request by AvivCX**
By submitting a PR to this repository, you agree to the terms within the [Checkmarx Code of Conduct](https://github.com/checkmarx-ts/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/checkmarx-ltd/open-source-template/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

Purpose of this PR is to add threshold logic for cxflow cli flows
Now user can configure thresholds to control break-build operations

### References

Documentation guide: https://github.com/checkmarx-ltd/cx-flow/wiki/Thresholds-and-policies

### Testing

Add automation voerage to cover cli flows in the following feature files:
- thresholds.feature
- sca-thresholds.feature

### Checklist

- [x] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentaiton is a Wiki Update, please indicate desired changes within PR MD Comment*
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java [hunk 1]
```diff
@@ -53,30 +54,58 @@ public ThresholdValidatorImpl(@Lazy SastScanner sastScanner, @Lazy SCAScanner sc
...
-            isAllowed = isAllowedSast(scanResults, pullRequestReport);
+            isAllowed = isAllowedSast(scanResults, request);
         }
 
         if (isAllowed && scaScanner.isEnabled()) {
-            isAllowed = isAllowedSca(scanResults, pullRequestReport);
+            isAllowed = isAllowedSca(scanResults, request);
         }
 
         return isAllowed;
```
Comment (ofersk) [inline src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:111]: this looks more readable to me:
```java
boolean isSastAllowed = isSast() ? isAllowedSast(scanResults, request) : true;
boolean isScaAllowed = scaScanner.isEnabled() ?  isAllowedSca(scanResults, request) : true;
return isSastAllowed && isScaAllowed 
```

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java [hunk 2]
```diff
@@ -507,6 +508,25 @@ private void processResults(ScanRequest request, ScanResults results) throws Exi
         }
     }
 
+    private boolean checkIfBreakBuild(ScanRequest request, ScanResults results){
+        boolean breakBuildResult = false;
+
+        if(thresholdValidator.isThresholdsConfigurationExist(request)){
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:514]: Actually this is different case.
If thresholds not exist I want to go check the break build, but if exist and exceeded, shouldn't go to the 'else' flow

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java [hunk 3]
```diff
@@ -507,6 +508,25 @@ private void processResults(ScanRequest request, ScanResults results) throws Exi
         }
     }
 
+    private boolean checkIfBreakBuild(ScanRequest request, ScanResults results){
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:511]: I don't think the ThresholdValidator should be aware of 'build', 'merge' or 'pull request' terminology
This is why I added 'isThresholdsConfigurationExist' as a generic check

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java [hunk 4]
```diff
@@ -53,30 +54,58 @@ public ThresholdValidatorImpl(@Lazy SastScanner sastScanner, @Lazy SCAScanner sc
...
+            log.info("Merge is allowed, because error-merge is set to false.");
+            pullRequestReport.setPullRequestResult(requestResult);
+        }
+        else{
+            isMergeAllowed = isAllowed(scanResults, pullRequestReport.getScanRequest());
+
+            if (!isMergeAllowed) {
+                log.info("Merge is not allowed, because some thresholds were exceeded.");
+                requestResult = new OperationResult(OperationStatus.FAILURE, MERGE_FAILURE_DESCRIPTION);
+            }
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:69]: Added log

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java (hunk 1 above)
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:111]: The result would be different. we want that if one of the scanner is enabled and its thresholds were exceeded - to return false (not allowed)
in your case, if SAST not enabled (only SCA), and SCA thresholds not exceeded - still we will return notAllowed

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java [hunk 5]
```diff
@@ -507,6 +508,25 @@ private void processResults(ScanRequest request, ScanResults results) throws Exi
...
+        boolean breakBuildResult = false;
+
+        if(thresholdValidator.isThresholdsConfigurationExist(request)){
+            if(thresholdValidator.thresholdsExceeded(request, results)){
+                log.info("Fail build on Thresholds exceeded.");
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:516]: it is logged inside the ThresholdValidator.
regarding 'if' statement see 1st Ofer comment

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java [hunk 6]
```diff
@@ -507,6 +508,25 @@ private void processResults(ScanRequest request, ScanResults results) throws Exi
...
+                breakBuildResult = true;
+            }
+        } else if(flowProperties.isBreakBuild() && resultsService.filteredIssuesPresent(results)){
+            log.info("Fail build on issues presented after executing filters");
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:520]: If thresholds not exist, the logic is to fail the build if there were left vulnerabilities after the processing 
https://github.com/checkmarx-ltd/cx-flow/wiki/Thresholds-and-policies#thresholds-vs-basic-filters

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java [hunk 7]
```diff
@@ -86,37 +115,34 @@ private boolean isSast() {
...
         isAllowedScannerToLog(isAllowed);
 
         return isAllowed;
     }
 
     private void isAllowedScannerToLog(boolean isAllowedScanner) {
-        log.info(isAllowedScanner ? "Merge is allowed, because no thresholds were exceeded." :
-                "Merge is not allowed, because some of the thresholds were exceeded.");
+        log.info(isAllowedScanner ? "No thresholds were exceeded." :
+                "Thresholds were exceeded.");
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:145]: ThresholdValidator check the thresholds logic, and shouldn't be aware of 'build', 'merge' or 'pull request'.
Moreover in CLI flow, the original message is wrong

File: src/main/java/com/checkmarx/flow/CxFlowRunner.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/CxFlowRunner.java:520]: changed

File: src/main/java/com/checkmarx/flow/service/ResultsService.java [hunk 8]
```diff
@@ -335,6 +335,11 @@ private boolean requiresCxCustomFields(List<Field> fields) {
     }
 
     public boolean filteredIssuesPresent(ScanResults results) {
+        if(results.getAdditionalDetails()==null) {
+            // assuming SCA results only
+            return false;
+        }
+
         Map<String, Integer> flow = (Map<String, Integer>) results.getAdditionalDetails().get(Constants.SUMMARY_KEY);
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ResultsService.java:343]: changed

File: src/main/java/com/checkmarx/flow/service/ResultsService.java [hunk 9]
```diff
@@ -335,6 +335,11 @@ private boolean requiresCxCustomFields(List<Field> fields) {
     }
 
     public boolean filteredIssuesPresent(ScanResults results) {
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ResultsService.java:337]: changed

File: src/main/java/com/checkmarx/flow/service/ResultsService.java [hunk 10]
```diff
@@ -335,6 +335,11 @@ private boolean requiresCxCustomFields(List<Field> fields) {
...
+        if(results.getAdditionalDetails()==null) {
+            // assuming SCA results only
+            return false;
+        }
+
         Map<String, Integer> flow = (Map<String, Integer>) results.getAdditionalDetails().get(Constants.SUMMARY_KEY);
         return flow == null || !flow.isEmpty();
```
[the remark this answers is not in the document]
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ResultsService.java:344]: in this case !flow.isEmpty() will return false (not present)

**Review by NimrodGolan — APPROVED** [review]

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java (hunk 1 above)
Comment (ofersk) [reply src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:115]: if SAST not enabled -> isSast() == false -> boolean isSastAllowed = true
SCA thresholds not exceeded -> scaScanner.isEnabled() == true -> isAllowedSca(scanResults, request) -> true
return isSastAllowed && isScaAllowed -> true && true -> true
still we will return notAllowed <- no as you can see we will return true

File: src/test/java/com/checkmarx/flow/cucumber/integration/cli/sast/SastCliSteps.java [hunk 11]
```diff
@@ -138,13 +144,22 @@ public void runnningScanWithFilter(String filter) {
...
+        Throwable exception = null;
         try {
             TestUtils.runCxFlow(testContext.getCxFlowRunner(), commandBuilder.toString());
         } catch (Throwable e) {
+            exception = e;
         }
+        testContext.setCxFlowExecutionException(exception);
+    }
+
+    @And("cxflow should exit with {}")
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/integration/cli/sast/SastCliSteps.java:156]: sould be {int} and not {}

File: src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java (hunk 1 above)
Comment (AvivCX) [reply src/main/java/com/checkmarx/flow/service/ThresholdValidatorImpl.java:115]: You are right I changed that (and simplified per intelij suggestion)

File: src/main/java/com/checkmarx/flow/service/ResultsService.java [hunk 12]
```diff
@@ -334,14 +334,14 @@ private boolean requiresCxCustomFields(List<Field> fields) {
...
+    public boolean filteredSastIssuesPresent(ScanResults results) {
         if(results.getAdditionalDetails()==null) {
             // assuming SCA results only
             return false;
         }
 
-        Map<String, Integer> flow = (Map<String, Integer>) results.getAdditionalDetails().get(Constants.SUMMARY_KEY);
-        return flow == null || !flow.isEmpty();
+        Map<String, Integer> findingCountPerSeverity = (Map<String, Integer>) results.getAdditionalDetails().get(Constants.SUMMARY_KEY);
+        return findingCountPerSeverity == null || !findingCountPerSeverity.isEmpty();
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/ResultsService.java:344]: If e.g. findingCountPerSeverity has a single 'HIGH' key with the value 0, then findingCountPerSeverity.isEmpty() will return false. The result of filteredSastIssuesPresent() will then be 'true'. However, in fact no issues are present.

**Pull request merged**